hoomd/mpcd/MPCDReversePerturbationFlow.cc

```diff
@@ -0,0 +1,339 @@
+    m_num_hi = 0;
+    for (auto iter = particles_in_hi_slab.begin(); iter != particles_in_hi_slab.end(); ++iter)
+        {
+        if (m_num_hi < m_num_swap)
+            {
+            h_layer_hi.data[m_num_hi++] = make_scalar2(__int_as_scalar(iter->second), iter->first);
+            }
+        }
+    }
+
+/*!
+ * Takes the up to m_num_swap pairs from the top and bottom slab and swaps
+ * their velocities.
+ */
+void mpcd::MPCDReversePerturbationFlow::swapPairMomentum()
+    {
+    // get all the needed array handles. Only velocities needs to be readwrite
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(),
+                               access_location::host,
+                               access_mode::readwrite);
+    ArrayHandle<Scalar2> h_layer_lo(m_layer_lo, access_location::host, access_mode::read);
+    ArrayHandle<Scalar2> h_layer_hi(m_layer_hi, access_location::host, access_mode::read);
+
+    // find out how many pairs there are
+    unsigned int num_pairs = std::min(m_num_swap, std::min(m_num_lo, m_num_hi));
+
+    // swap num_pairs (up to m_num_swap)
+    m_momentum_exchange = 0;
+    for (unsigned int i = 0; i < num_pairs; i++)
+        {
+        // load data from the lo slab
+        const Scalar2 lo = h_layer_lo.data[i];
+        const unsigned int lo_idx = __scalar_as_int(lo.x);
+        const Scalar lo_momentum = lo.y;
+
+        // load data from the hi slab
+        const Scalar2 hi = h_layer_hi.data[i];
+        const unsigned int hi_idx = __scalar_as_int(hi.x);
+        const Scalar hi_momentum = hi.y;
+        // swap velocities & calculate momentum exchange
+        m_momentum_exchange += (lo_momentum - hi_momentum);
+        h_vel.data[lo_idx].x = hi_momentum / m_mpcd_pdata->getMass();
+        h_vel.data[hi_idx].x = lo_momentum / m_mpcd_pdata->getMass();
+        }
+    }
+
+// mask here: namespace detail
+// and this: {
+/*!
+ * \param m Python module to export to
+ */
+void mpcd::detail::export_MPCDReversePerturbationFlow(pybind11::module& m)
+    {
+    namespace py = pybind11;
+    py::class_<mpcd::MPCDReversePerturbationFlow,
+               std::shared_ptr<mpcd::MPCDReversePerturbationFlow>>(m,
+                                                                   "MPCDReversePerturbationFlow",
+                                                                   py::base<Updater>())
+        .def(py::init<std::shared_ptr<SystemDefinition>,
+                      std::shared_ptr<Trigger>,
+                      unsigned int,
+                      Scalar,
+                      Scalar,
+                      Scalar>())
+        .def_property("Nswap",
+                      &mpcd::MPCDReversePerturbationFlow::getNswap,
+                      &mpcd::MPCDReversePerturbationFlow::setNswap)
+        .def_property("width",
+                      &mpcd::MPCDReversePerturbationFlow::getSlabWidth,
+                      &mpcd::MPCDReversePerturbationFlow::setSlabWidth)
+        .def_property("distance",
```

Comment:
For consistency with some other geometries, I would probably call this `slab_separation`
```suggestion
        .def_property("slab_separation",
```